contracts/contracts/strategies/GeneralizedUniswapV3Strategy.sol

```diff
@@ -0,0 +1,850 @@
+    /***************************************
+            Deposit/Withdraw
+    ****************************************/
+
+    /// @inheritdoc InitializableAbstractStrategy
+    function deposit(address _asset, uint256 _amount)
+        external
+        override
+        onlyVault
+        nonReentrant
+    {
+        require(_asset == token0 || _asset == token1, "Unsupported asset");
+        IVault(vaultAddress).depositForUniswapV3(_asset, _amount);
+        // Not emitting Deposit event since the Reserve strategy would do so
+    }
+
+    /// @inheritdoc InitializableAbstractStrategy
+    function depositAll() external override onlyVault nonReentrant {
+        _depositAll();
+    }
+
+    /**
+     * @notice Deposits all undeployed balances of the contract to the reserve strategies
+     */
+    function _depositAll() internal {
+        uint256 token0Bal = IERC20(token0).balanceOf(address(this));
+        uint256 token1Bal = IERC20(token1).balanceOf(address(this));
+        if (token0Bal > 0) {
+            IVault(vaultAddress).depositForUniswapV3(token0, token0Bal);
+        }
+        if (token1Bal > 0) {
+            IVault(vaultAddress).depositForUniswapV3(token1, token1Bal);
+        }
+
+        // Not emitting Deposit events since the Reserve strategies would do so
+    }
+
+    /**
+     * @notice Withdraws asset from the reserve strategy
+     * @inheritdoc InitializableAbstractStrategy
+     */
+    function withdraw(
+        address recipient,
+        address _asset,
+        uint256 amount
+    ) external override onlyVault nonReentrant {
+        require(_asset == token0 || _asset == token1, "Unsupported asset");
+
+        IERC20 asset = IERC20(_asset);
+        uint256 selfBalance = asset.balanceOf(address(this));
+
+        if (selfBalance < amount) {
+            // Try to pull remaining amount from reserve strategy
+            // This might throw if there isn't enough in reserve strategy as well
+            IVault(vaultAddress).withdrawForUniswapV3(
+                recipient,
+                _asset,
+                amount - selfBalance
+            );
+
+            // TODO: Remove liquidity from V3 pool instead?
+
+            // Transfer all of unused balance
+            asset.safeTransfer(recipient, selfBalance);
+
+            // Emit event for only the amount transferred out from this strategy
+            emit Withdrawal(_asset, _asset, selfBalance);
+        } else {
+            // Transfer requested amount
+            asset.safeTransfer(recipient, amount);
+            emit Withdrawal(_asset, _asset, amount);
+        }
+    }
+
+    /**
+     * @notice Closes active LP position, if any, and transfer all token balance to Vault
+     * @inheritdoc InitializableAbstractStrategy
+     */
+    function withdrawAll() external override onlyVault nonReentrant {
+        if (currentPositionTokenId > 0) {
+            _closePosition(currentPositionTokenId);
+        }
+
+        IERC20 cToken0 = IERC20(token0);
```

Comment:
These aren't ctokens? should this just be token0 and token1? We should probably fix throughout the code.

Reply:
I used the names `token0` and `token1` for the addresses. Since this is an instance of the contract, I was thinking `Contract of Token 0 and 1` when I named it `cToken0`. But I agree it's confusing with Compound Tokens, will change it